hal/src/sercom/v2/spi.rs

```diff
@@ -0,0 +1,1348 @@
+    fn as_ref(&self) -> &Self {
+        self
+    }
+}
+
+impl<P, M, Z> AsMut<Self> for Config<P, M, Z>
+where
+    P: ValidPads,
+    M: OpMode,
+    Z: Size,
+{
+    #[inline]
+    fn as_mut(&mut self) -> &mut Self {
+        self
+    }
+}
+
+//=============================================================================
+// ValidConfig
+//=============================================================================
+
+/// Marker trait for valid SPI [`Config`]urations
+///
+/// A functional SPI peripheral must have, at a minimum, an SCLK pad and
+/// either a Data In or a Data Out pad. Dependeing on the [`OpMode`], an SS
+/// pad may also be required.
+///
+/// The `ValidConfig` trait is implemented only for valid combinations of
+/// [`Pads`] and [`OpMode`]. No [`Config`] is valid if the SCK pad is [`NoneT`]
+/// or if both the Data In and Data Out pads are `NoneT`. When in [`Master`]
+/// `OpMode`, the `SS` pad must be `NoneT`, while in [`MasterHWSS`] or
+/// [`Slave`] [`OpMode`], the SS pad must be [`SomePad`].
+pub trait ValidConfig: AnyConfig {}
+
+impl<P, Z> ValidConfig for Config<P, Master, Z>
+where
+    P: ValidPads<SS = NoneT>,
+    Z: Size,
+{
+}
+
+impl<P, Z> ValidConfig for Config<P, MasterHWSS, Z>
+where
+    P: ValidPads,
+    Z: Size,
+    P::SS: SomePad,
+{
+}
+
+impl<P, Z> ValidConfig for Config<P, Slave, Z>
+where
+    P: ValidPads,
+    Z: Size,
+    P::SS: SomePad,
+{
+}
+
+//=============================================================================
+// Spi
+//=============================================================================
+
+/// An enabled SPI peripheral that can perform transactions
+///
+/// See the [`impl_ehal`] documentation for details on the implementations of
+/// the embedded HAL traits, which vary based on [`Size`] and [`Capability`].
+pub struct Spi<C, A>
+where
+    C: ValidConfig,
+    A: Capability,
+{
+    config: C,
+    capability: A,
+}
+
+/// Get a shared reference to the underlying [`Config`] struct
+///
+/// This can be used to call the various `get_*` functions on `Config`
+impl<C, A> AsRef<SpecificConfig<C>> for Spi<C, A>
+where
+    C: ValidConfig,
+    A: Capability,
+{
+    #[inline]
+    fn as_ref(&self) -> &SpecificConfig<C> {
+        self.config.as_ref()
+    }
+}
+
+impl<C, A> Spi<C, A>
+where
+    C: ValidConfig,
+    A: Capability,
+{
+    /// Change the transaction [`Length`]
+    ///
+    /// Changing the transaction [`Length`] while is enabled is permissible but
+    /// `unsafe`. If you have sent or received *any* bytes at the current
+    /// [`Length`], you **must** wait for a TXC flag before changing to a new
+    /// [`Length`].
+    #[inline]
+    #[cfg(feature = "min-samd51g")]
+    pub unsafe fn length<L: Length>(self) -> Spi<Config<C::Pads, C::OpMode, L>, A>
+    where
+        Config<C::Pads, C::OpMode, L>: ValidConfig,
+    {
+        Spi {
+            config: self.config.into().length(),
+            capability: self.capability,
+        }
+    }
```

Comment:
Is it really `unsafe`? In a previous discussion we agreed that `unsafe` should, as much as possible, be used only for things that may incur memory unsafety/UB. I guess it could be argued that this is "memory unsafe" as the data received is not what was expected. I'm not sure.